src/dmd/parse.d

```diff
@@ -6211,7 +6211,16 @@ LagainStc:
                     auto statements = new AST.Statements();
                     while (token.value != TOK.case_ && token.value != TOK.default_ && token.value != TOK.endOfFile && token.value != TOK.rightCurly)
                     {
-                        statements.push(parseStatement(ParseStatementFlags.semi | ParseStatementFlags.curlyScope));
+                        auto cur = parseStatement(ParseStatementFlags.semi | ParseStatementFlags.curlyScope);
+                        statements.push(cur);
+
+                        // https://issues.dlang.org/show_bug.cgi?id=21739
+                        // Stop at the last break s.t. the following non-case statements are
+                        // not merged into the current case. This can happen for
+                        // case 1: ... break;
+                        // debug { case 2: ... }
+                        if (cur.isBreakStatement())
```

Comment:
Isn't this solution going to detach the `case` inside the `debug` statement from the switch? For example, does this valid code still works?

```d
void main()
{
    int i, k;
    switch (i)
    {
        case 0:
            int x;
            break;

        case 1:
            x = 1;
            break;

        case 2:
            int y;
            break;

        debug
        {
            case 3:
                k = 1; // Accepted
                break;
        }

        default:
    }
}
```
Anyway, this test should probably added as a compilable test.

Reply:
No, it just splits the `case 3` from the previous case.

Extended the tests according to your example.